presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoResultSet.java

```diff
@@ -331,16 +326,28 @@ private Timestamp getTimestamp(int columnIndex, DateTimeZone localTimeZone)
             }
         }
 
+        throw new IllegalArgumentException("Expected column to be a timestamp type but is " + columnInfo.getColumnTypeName());
+    }
+
+    private ZonedDateTime getTimestampWithTimeZone(int columnIndex)
+            throws SQLException
+    {
+        Object value = column(columnIndex);
+        if (value == null) {
+            return null;
+        }
+
+        ColumnInfo columnInfo = columnInfo(columnIndex);
         if (columnInfo.getColumnTypeName().equalsIgnoreCase("timestamp with time zone")) {
             try {
-                return new Timestamp(TIMESTAMP_WITH_TIME_ZONE_FORMATTER.parseMillis(String.valueOf(value)));
+                return TIMESTAMP_WITH_TIME_ZONE_FORMATTER.parse(String.valueOf(value), ZonedDateTime::from);
```

Comment:
@dain good question. As discussed offline, this pertains to `ResultSet#getObject(col)` (i.e. without `Class` specifier).
From the database I have some working knowledge of, Oracle implements TIMESTAMP WITH TIME ZONE correctly, and probably Teradata as well.
I don't know yet what Oracle driver returns on `getObject(col)` (i would assume something compatible with what they picked when implementing JDBC 3...)

I think in Presto, `getObject(col)` should *not* return `OffsetDateTime`. This would be misleading, as the underlying value is not representable as  `OffsetDateTime` without information loss.

Reply:
PostgreSQL JDBC `42.2.5` returns `java.sql.Timestamp` by default and supports `OffsetDateTime` but not `ZonedDateTime`.

Reply:
The PostgreSQL [docs](https://www.postgresql.org/docs/11/datatype-datetime.html#id-1.5.7.13.19.7):

> For timestamp with time zone, the internally stored value is always in UTC

> When a timestamp with time zone value is output, it is always converted from UTC to the current timezone zone, and displayed as local time in that zone

This seems to violate the SQL standard (and is different than Oracle), and the documentation doesn't mention it (though it mentions the standard in several other places).

Reply:
Right, PostgreSQL's`TIMESTAMP WITH TIME ZONE` is not standard-compliant. Their type represents a point in time (Instant in Java) instead of (point in time, zone) pair (ZonedDateTime in Java).
Thus, i don't see need to align with PostgreSQL driver behavior for this feature.

Reply:
Agreed, returning `ZonedDateTime` seems correct for Presto.